lib/charms/operator_libs_linux/v1/snap.py

```diff
@@ -262,21 +264,35 @@ def __str__(self):
             str(self._state),
         )
 
-    def _snap(self, command: str, optargs: Optional[Iterable[str]] = None) -> str:
+    @retry(
+        stop=stop_after_attempt(5),
+        wait=wait_fixed(5),
+        reraise=True,
+        before=before_log(logger, logging.DEBUG),
+    )
```

Comment:
My concern with adding retries is that we may be somewhat blindly patching over an issue that we don't know the cause of.

I think part of the issue is we're not showing the stderr from the `subprocess.check_output()` call, only the stderr, which isn't very useful. The error information from snap will be in stderr, and stdout will be empty. So I think we should change the `check_output()` block below to the following:

```python
try:
    return subprocess.check_output(_cmd, universal_newlines=True, stderr=subprocess.PIPE)
except CalledProcessError as e:
    raise SnapError(
        "Snap: {!r}; command {!r} failed with stderr = {!r}".format(
            self._name, _cmd, e.stderr
        )
    )
```

The above change should also be done in the `check_output()` call in `install_local` and its exception handling, and in `_system_set` below that:

```python
    try:
        subprocess.check_call(_cmd, universal_newlines=True, stderr=subprocess.PIPE)
    except CalledProcessError as e:
        raise SnapError("Failed setting system config '{}' to '{}': {}".format(config_item, value, e.stderr))
```

I don't think we should add the timeout or the retries at this point, until we know more.

Reply:
Would be good to add unit tests that check the error case and assert that the stderr is in the exception message. Maybe by patching `subprocess.check_output` with a version that instead calls `check_output(['sleep', 'x'])` or something, so it's slightly realistic?

Reply:
We could remove `timeout` for the time being until we figure out what the issue is here. (But, debugging this will take significant time, especially considering that without the `timeout` it can take up to several hours for `subprocess.check_call` to finish.)

 I have limited time and this issue is a blocker for my team. I will see if I have time to debug this issue.